# Cell 1 — Import Libraries

## What I'm Doing

In this cell, I'm importing all the libraries I'll need throughout this notebook. Nothing runs here — I'm just loading the tools upfront so everything is ready when I need it later.

---

## Libraries I'm Bringing In

### Data Handling
- **pandas** — for loading, cleaning, and manipulating my dataset
- **numpy** — for numerical operations and array handling

### Visualization
- **matplotlib** — for plotting charts and graphs
- **seaborn** — for cleaner, more polished statistical visualizations

### Classical Machine Learning (scikit-learn)
- **train_test_split** — to split my data into training and testing sets
- **StandardScaler** — to normalize my features before feeding them into a model
- **LogisticRegression** — for classification tasks
- **LinearRegression** — for regression tasks
- **Metrics** — to evaluate model performance (MAE, MSE, R², confusion matrix, classification report)

### Deep Learning
- **PyTorch (torch)** — the core deep learning framework I'll use to build a neural network
- **torch.nn** — gives me access to layers, activations, and loss functions
- **DataLoader & TensorDataset** — for batching and feeding data efficiently into the neural network during training

---

## End-of-Cell Checks

At the end of the cell, I'm running three quick checks:

1. Confirming everything imported without errors
2. Printing the PyTorch version I'm running
3. Checking whether a GPU is available — if `True`, training will be significantly faster

---

## Summary

This cell is purely setup. Once it runs cleanly, I'm ready to move on to loading data and building models.

In [2]:
# ============================================================
# Cell 1: Import Libraries
# All the tools we need for data handling, modeling, and plots
# ============================================================

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression, LinearRegression
from sklearn.metrics import (
    classification_report,
    confusion_matrix,
    mean_absolute_error,
    mean_squared_error,
    r2_score
)

import torch
import torch.nn as nn
from torch.utils.data import DataLoader, TensorDataset

print("All libraries imported successfully.")
print("PyTorch version:", torch.__version__)
print("GPU available:", torch.cuda.is_available())

All libraries imported successfully.
PyTorch version: 2.10.0+cu128
GPU available: True


All libraries loaded successfully. We are running PyTorch 2.10 with CUDA available, confirming the A100 GPU is active and ready. This means our neural networks will train on GPU, which will significantly speed up the training process compared to CPU.

# Cell 2 — Download and Load the Dataset

## What I'm Doing

In this cell, I'm pulling in the dataset I'll be working with and loading it into memory so I can start exploring it.

---

## Step by Step

### Importing kagglehub
The UCI server where this dataset originally lives was unavailable, so instead I'm using **kagglehub** — a Python package that lets me download datasets directly from Kaggle into my environment.

### Downloading the Dataset
I'm downloading the **Diabetes Health Indicators Dataset** from Kaggle. Once it's downloaded, kagglehub returns the local path where the files were saved, and I'm printing that path just to confirm where it landed.

### Loading it into a DataFrame
I'm reading the specific CSV file from that downloaded path into a **pandas DataFrame** called `df`. This is now my main working object — everything from here on will be done on `df`.

### Quick Sanity Checks
At the end I'm running two checks:

- **Shape** — tells me how many rows and columns the dataset has, so I know the scale of what I'm working with
- **df.head()** — shows me the first 5 rows so I can visually confirm the data loaded correctly and get a feel for what the columns look like

---

## Summary

By the end of this cell, my dataset is downloaded, loaded, and I've done a first glance at it. I'm ready to start understanding and cleaning it in the next steps.

In [3]:
# ============================================================
# Cell 2: Download and Load the Dataset
# UCI server was unavailable so we use kagglehub to fetch
# the same dataset directly into the Colab environment
# ============================================================

import kagglehub

path = kagglehub.dataset_download("alexteboul/diabetes-health-indicators-dataset")
print("Dataset downloaded to:", path)

df = pd.read_csv(path + "/diabetes_binary_health_indicators_BRFSS2015.csv")

print("Shape:", df.shape)
print("\nFirst 5 rows:")
df.head()

100%|██████████| 6.03M/6.03M [00:02<00:00, 2.95MB/s]

Extracting files...


Dataset downloaded to: /root/.cache/kagglehub/datasets/alexteboul/diabetes-health-indicators-dataset/versions/1
Shape: (253680, 22)

First 5 rows:


,Diabetes_binary,HighBP,HighChol,CholCheck,BMI,Smoker,Stroke,HeartDiseaseorAttack,PhysActivity,Fruits,...,AnyHealthcare,NoDocbcCost,GenHlth,MentHlth,PhysHlth,DiffWalk,Sex,Age,Education,Income
0,0.0,1.0,1.0,1.0,40.0,1.0,0.0,0.0,0.0,0.0,...,1.0,0.0,5.0,18.0,15.0,1.0,0.0,9.0,4.0,3.0
1,0.0,0.0,0.0,0.0,25.0,1.0,0.0,0.0,1.0,0.0,...,0.0,1.0,3.0,0.0,0.0,0.0,0.0,7.0,6.0,1.0
2,0.0,1.0,1.0,1.0,28.0,0.0,0.0,0.0,0.0,1.0,...,1.0,1.0,5.0,30.0,30.0,1.0,0.0,9.0,4.0,8.0
3,0.0,1.0,0.0,1.0,27.0,0.0,0.0,0.0,1.0,1.0,...,1.0,0.0,2.0,0.0,0.0,0.0,0.0,11.0,3.0,6.0
4,0.0,1.0,1.0,1.0,24.0,0.0,0.0,0.0,1.0,1.0,...,1.0,0.0,2.0,3.0,0.0,0.0,0.0,11.0,5.0,4.0
